### ISSUE TITLE
Enhancement idea:  flags for territories

### ISSUE THREAD

**Issue by kipcole9** [Schultzer/cldr_territories#3]
@Schultzer Had a spare moment on a plane and thought maybe adding unicode flags to cldr_territories might be worth while (I know I have a couple of use cases).  I created a simple module below, happy to do a PR if that suits you.

```elixir
defmodule Cldr.Territory do
  @flags File.read!("priv/flags.csv")
  Enum.each String.split(@flags, "\n"), fn line ->
    case String.split(line, ",") do
      [flag, territory] ->
        def flag_for(unquote(String.to_atom(territory))), do: {:ok, unquote(String.trim(flag))}
      [_] ->
        nil
    end
  end

  def flag_for(territory) do
    {:error, {Cldr.UnknownTerritoryError, "Unknown territory #{inspect territory}"}}
  end

  def flag_for!(territory) do
    case flag_for(territory) do
      {:ok, flag} -> flag
      {:error, {exception, reason}} -> raise exception, reason
    end
  end
end
```

The flags file is nothing more than:
```
🇦🇨,AC
🇦🇩,AD
🇦🇪,AE
🇦🇫,AF
🇦🇬,AG
🇦🇮,AI
🇦🇱,AL
🇦🇲,AM
🇦🇴,AO
🇦🇶,AQ
🇦🇷,AR
🇦🇸,AS
🇦🇹,AT
🇦🇺,AU
🇦🇼,AW
🇦🇽,AX
🇦🇿,AZ
🇧🇦,BA
🇧🇧,BB
🇧🇩,BD
🇧🇪,BE
🇧🇫,BF
🇧🇬,BG
🇧🇭,BH
🇧🇮,BI
🇧🇯,BJ
🇧🇱,BL
🇧🇲,BM
🇧🇳,BN
🇧🇴,BO
🇧🇶,BQ
🇧🇷,BR
🇧🇸,BS
🇧🇹,BT
🇧🇻,BV
🇧🇼,BW
🇧🇾,BY
🇧🇿,BZ
🇨🇦,CA
🇨🇨,CC
🇨🇩,CD
🇨🇫,CF
🇨🇬,CG
🇨🇭,CH
🇨🇮,CI
🇨🇰,CK
🇨🇱,CL
🇨🇲,CM
🇨🇳,CN
🇨🇴,CO
🇨🇵,CP
🇨🇷,CR
🇨🇺,CU
🇨🇻,CV
🇨🇼,CW
🇨🇽,CX
🇨🇾,CY
🇨🇿,CZ
🇩🇪,DE
🇩🇬,DG
🇩🇯,DJ
🇩🇰,DK
🇩🇲,DM
🇩🇴,DO
🇩🇿,DZ
🇪🇦,EA
🇪🇨,EC
🇪🇪,EE
🇪🇬,EG
🇪🇭,EH
🇪🇷,ER
🇪🇸,ES
🇪🇹,ET
🇪🇺,EU
🇫🇮,FI
🇫🇯,FJ
🇫🇰,FK
🇫🇲,FM
🇫🇴,FO
🇫🇷,FR
🇬🇦,GA
🇬🇧,GB
🇬🇩,GD
🇬🇪,GE
🇬🇫,GF
🇬🇬,GG
🇬🇭,GH
🇬🇮,GI
🇬🇱,GL
🇬🇲,GM
🇬🇳,GN
🇬🇵,GP
🇬🇶,GQ
🇬🇷,GR
🇬🇸,GS
🇬🇹,GT
🇬🇺,GU
🇬🇼,GW
🇬🇾,GY
🇭🇰,HK
🇭🇲,HM
🇭🇳,HN
🇭🇷,HR
🇭🇹,HT
🇭🇺,HU
🇮🇨,IC
🇮🇩,ID
🇮🇪,IE
🇮🇱,IL
🇮🇲,IM
🇮🇳,IN
🇮🇴,IO
🇮🇶,IQ
🇮🇷,IR
🇮🇸,IS
🇮🇹,IT
🇯🇪,JE
🇯🇲,JM
🇯🇴,JO
🇯🇵,JP
🇰🇪,KE
🇰🇬,KG
🇰🇭,KH
🇰🇮,KI
🇰🇲,KM
🇰🇳,KN
🇰🇵,KP
🇰🇷,KR
🇰🇼,KW
🇰🇾,KY
🇰🇿,KZ
🇱🇦,LA
🇱🇧,LB
🇱🇨,LC
🇱🇮,LI
🇱🇰,LK
🇱🇷,LR
🇱🇸,LS
🇱🇹,LT
🇱🇺,LU
🇱🇻,LV
🇱🇾,LY
🇲🇦,MA
🇲🇨,MC
🇲🇩,MD
🇲🇪,ME
🇲🇫,MF
🇲🇬,MG
🇲🇭,MH
🇲🇰,MK
🇲🇱,ML
🇲🇲,MM
🇲🇳,MN
🇲🇴,MO
🇲🇵,MP
🇲🇶,MQ
🇲🇷,MR
🇲🇸,MS
🇲🇹,MT
🇲🇺,MU
🇲🇻,MV
🇲🇼,MW
🇲🇽,MX
🇲🇾,MY
🇲🇿,MZ
🇳🇦,NA
🇳🇨,NC
🇳🇪,NE
🇳🇫,NF
🇳🇬,NG
🇳🇮,NI
🇳🇱,NL
🇳🇴,NO
🇳🇵,NP
🇳🇷,NR
🇳🇺,NU
🇳🇿,NZ
🇴🇲,OM
🇵🇦,PA
🇵🇪,PE
🇵🇫,PF
🇵🇬,PG
🇵🇭,PH
🇵🇰,PK
🇵🇱,PL
🇵🇲,PM
🇵🇳,PN
🇵🇷,PR
🇵🇸,PS
🇵🇹,PT
🇵🇼,PW
🇵🇾,PY
🇶🇦,QA
🇷🇪,RE
🇷🇴,RO
🇷🇸,RS
🇷🇺,RU
🇷🇼,RW
🇸🇦,SA
🇸🇧,SB
🇸🇨,SC
🇸🇩,SD
🇸🇪,SE
🇸🇬,SG
🇸🇭,SH
🇸🇮,SI
🇸🇯,SJ
🇸🇰,SK
🇸🇱,SL
🇸🇲,SM
🇸🇳,SN
🇸🇴,SO
🇸🇷,SR
🇸🇸,SS
🇸🇹,ST
🇸🇻,SV
🇸🇽,SX
🇸🇾,SY
🇸🇿,SZ
🇹🇦,TA
🇹🇨,TC
🇹🇩,TD
🇹🇫,TF
🇹🇬,TG
🇹🇭,TH
🇹🇯,TJ
🇹🇰,TK
🇹🇱,TL
🇹🇲,TM
🇹🇳,TN
🇹🇴,TO
🇹🇷,TR
🇹🇹,TT
🇹🇻,TV
🇹🇼,TW
🇹🇿,TZ
🇺🇦,UA
🇺🇬,UG
🇺🇲,UM
🇺🇳,UN
🇺🇸,US
🇺🇾,UY
🇺🇿,UZ
🇻🇦,VA
🇻🇨,VC
🇻🇪,VE
🇻🇬,VG
🇻🇮,VI
🇻🇳,VN
🇻🇺,VU
🇼🇫,WF
🇼🇸,WS
🇽🇰,XK
🇾🇪,YE
🇾🇹,YT
🇿🇦,ZA
🇿🇲,ZM
🇿🇼,ZW
```

**Comment by Schultzer** [comment]
@kipcole9, first of all, I think this is a really good idea.

But I would properly do something like this 

let me know if this fits your use case? 

if you like this direction I will implement this with fallback and test.
```elixir
  @spec to_unicode_flag(atom) :: String.t
  def to_unicode_flag(territory_code) do
    territory_code
    |> Atom.to_charlist
    |> Enum.map(fn x -> binary_to_codepoint(<<x>>) end)
    |> List.to_string
  end
  
  def binary_to_codepoint(<<65>>), do: [127462]
  def binary_to_codepoint(<<66>>), do: [127463]
  def binary_to_codepoint(<<67>>), do: [127464]
  def binary_to_codepoint(<<68>>), do: [127465]
  def binary_to_codepoint(<<69>>), do: [127466]
  def binary_to_codepoint(<<70>>), do: [127467]
  def binary_to_codepoint(<<71>>), do: [127468]
  def binary_to_codepoint(<<72>>), do: [127469]
  def binary_to_codepoint(<<73>>), do: [127470]
  def binary_to_codepoint(<<74>>), do: [127471]
  def binary_to_codepoint(<<75>>), do: [127472]
  def binary_to_codepoint(<<76>>), do: [127473]
  def binary_to_codepoint(<<77>>), do: [127474]
  def binary_to_codepoint(<<78>>), do: [127475]
  def binary_to_codepoint(<<79>>), do: [127476]
  def binary_to_codepoint(<<80>>), do: [127477]
  def binary_to_codepoint(<<81>>), do: [127478]
  def binary_to_codepoint(<<82>>), do: [127479]
  def binary_to_codepoint(<<83>>), do: [127480]
  def binary_to_codepoint(<<84>>), do: [127481]
  def binary_to_codepoint(<<85>>), do: [127482]
  def binary_to_codepoint(<<86>>), do: [127483]
  def binary_to_codepoint(<<87>>), do: [127484]
  def binary_to_codepoint(<<88>>), do: [127485]
  def binary_to_codepoint(<<89>>), do: [127486]
  def binary_to_codepoint(<<90>>), do: [127487]  
```

**Comment by kipcole9** [comment]
Yes, thats much nicer.  Perhaps wrap the api call with `validate_territory/1` to be consistent with the rest of the api?  Also, could collapse the function reference in the `Enum.map` call?

```elixir
  @spec to_unicode_flag(atom) :: String.t
  def to_unicode_flag(territory_code) do
    with {:ok, territory_code} <- Cldr.validate_territory(territory_code) do
      territory_code
      |> Atom.to_charlist
      |> Enum.map(&binary_to_codepoint/1)
      |> List.to_string
    else
      {:error, reason} -> {:error, reason}
    end
end
```
I haven't checked if the Unicode flag data includes all of the Cldr territories. Also would need a clean fallback for valid territories that have no flag, like `001`.

**Comment by Schultzer** [comment]
Okay sounds good :) I'm gonna do some more work on the guarding and tests. 
we can only construct flags for territory_codes that contain letters from A..Z
I'm planning to have this done today.

**Comment by kipcole9** [comment]
I’ve just put ex_cldr v1.0.0-rc.0 up on hex so perhaps thats the right development target?  I’m working on updating my other dependencies now to the same status (it’ll take a few days I think).

**Comment by Schultzer** [comment]
@kipcole9 
I'm closing this since [`f4570b1`](https://github.com/Schultzer/cldr_territories/commit/f4570b1466e1fe87972907673c5742c83fe74e95) fixes this issue

I've released a [`v1.0.0-rc.0`](https://hex.pm/packages/ex_cldr_territories) on hex.

let me know if you have any issues.
